src/lj_vmevent_printer.c

```diff
@@ -0,0 +1,262 @@
+#define LUA_CORE
+
+#include "lj_jit.h"
+#include "lj_trace.h"
+#include "lj_gc.h"
+#include "lj_vmevent.h"
+#include "lj_vmevent_printer.h"
+#include "lj_debug.h"
+#include "luajit.h"
+#include "lauxlib.h"
+
+typedef struct VMPrintContext {
+  VMEventContext user;
+  GCfunc *startfunc;
+  GCproto *lastpt;
+  BCPos lastpc;
+  GCfunc *lastlua;
+  GCfunc *lastfunc;
+} VMPrintContext;
+
+static uint32_t getlastline(VMPrintContext *context, GCproto *pt, BCPos pc)
+{
+  if(!pt) {
+    pt = funcproto(context->lastlua);
+    pc = context->lastpc;
+  }
+
+  return lj_debug_line(pt, pc);
+}
+
+static int isloopbc(BCOp op)
+{
+  return op == BC_LOOP || op == BC_FORI  || op == BC_FORL || op == BC_ITERN;
+}
+
+static void vmevent_tracestart(VMPrintContext *context, jit_State *J, void *eventdata)
+{
+  GCtrace *T = (GCtrace*)eventdata;
+  BCPos pc = J->pt ? (int32_t)proto_bcpos(J->pt, J->pc) : -1;
+  GCproto *pt = J->pt;
+  lua_assert(J->pt);
+
+  if (context->user.filter & EVENT_CLASS_TRACE) {
+    return;
+  }
+  
+  context->startfunc = J->fn;
+  context->lastfunc = context->lastlua = J->fn;
+
+  const char *start = "root";
+  if (J->pc && isloopbc(bc_op(*J->pc))) {
+    start = "loop";
+  }
+
+  if (J->parent == 0) {
+    fprintf(context->user.output, "START(%d): %s %s %d\n", T->traceno, start,  proto_chunknamestr(pt), lj_debug_line(pt, pc));
+  } else {
+    fprintf(context->user.output, "START(%d): side %s %d\n", T->traceno, proto_chunknamestr(pt), lj_debug_line(pt, pc));
+  }
+}
+
+static void vmevent_tracestop(VMPrintContext *context, jit_State *J, GCtrace *T)
+{
+  GCproto *pt = J->pt;
+  BCPos pc;
+
+  if (context->user.filter & EVENT_CLASS_TRACE) {
+    return;
+  }
+
+  if (!pt) {
+    pt = funcproto(context->lastlua);
+    pc = context->lastpc;
+  } else {
+    pc = proto_bcpos(pt, J->pc);
+  }
+
+  if (J->cur.linktype == LJ_TRLINK_INTERP) {
+    fprintf(context->user.output, "STOP(%d): Compiled exit to %s %d\n", T->traceno, proto_chunknamestr(pt), getlastline(context, pt, pc));
+  } else {
+    fprintf(context->user.output, "STOP(%d): %s %d\n", T->traceno, proto_chunknamestr(pt), getlastline(context, pt, pc));
+  }
+}
+
+static void vmevent_tracebc(VMPrintContext *context, jit_State *J, void *eventdata)
+{
+  GCproto *pt = J->pt;
+  BCPos pc = J->pt ? (int32_t)proto_bcpos(J->pt, J->pc) : -1;
+
+  if (context->lastfunc != J->fn) {
+    context->lastfunc = J->fn;
+  }
+
+  if (pt && J->fn != context->lastlua) {
+    lua_assert(pc != -1 && isluafunc(J->fn));
+    context->lastlua = J->fn;
+  }
+
+  if (pc != -1) {
+    context->lastpc = pc;
+    context->lastpt = pt;
+  }
+}
+
+static void vmevent_exit(VMPrintContext *context, jit_State *J, void *exitState)
+{
+  int gcexit = J2G(J)->gc.state == GCSatomic || J2G(J)->gc.state == GCSfinalize;
```

Comment:
I had separated it out as a variable because the meaning was not obvious and it would of made the fprintf less readable if it was written inline. I renamed the variable and added a comment